### PR TITLE
[WIP] Read csv can cheaply infere divisions

### PR DESCRIPTION
**Pull request by abast**
If an index column that is guaranteed to be sorted is provided to `dd.read_csv`, it should be possible to read in the divisions cheaply by sampling one data row at the beginning of each byte block. To make this possible, I extended my previous pull request (https://github.com/dask/dask/pull/1852) with the following (very experimental) changes:

- read_block in bytes/utils.py now has the keyword `number_of_delimiters`, which specifies the number of lines that should be sampled. This allows to read in e.g. the head and the first data row.
- read_bytes in bytes/core.py now provides the keyword `samples_per_block`. If it is set to `True`, a sample of the header + first data row of every block is returned in a list. If it is set to `False` (default), everything is as it was before. 
- read_pandas in dask/dataframe/io/csv.py provides the keywords `sorted_index_col` and `divisions`. It tries to calculate the divisions, if a sorted index column but no divisions are provided. To do this, it calls read_bytes with samples_per_block = True.

Might this be a good idea? Are there heavy pitfalls I am not aware of? I am new to github, so please feel free to correct me where ever you want.

### CODE REVIEW

**Comment by abast** [comment]
I did some benchmarks using my day to day data (2000 x 13MB csv files): 

```
#Read in without index and divisions (usual way)
%timeit -o dd.read_csv('[path]/dask_to_csv.*.csv')
> 1 loop, best of 3: 2.68 s per loop

#Read in with sorted_index_col = 'my_index'. Divisions will be calculated.
%timeit -o dd.read_csv('[path]/dask_to_csv.*.csv', sorted_index_col = 'my_index')
> 1 loop, best of 3: 8.41 s per loop

#Read in with sorted_index_col = 'my_index', blocksize = None and divisions = 'my_known_divisions'
%timeit -o dd.read_csv('[path]/dask_to_csv.*.csv', sorted_index_col = 'sim_trail_index', blocksize = None, divisions = divisions)
> 1 loop, best of 3: 2.66 s per loop
```

This is rapidly fast compared to the currently recommendet approach to use `ddf.set_index()`,  which I tried on a computesever with 40 cores:
```
#Read in without index and divisions and call set_index:
with dask.set_options(get = dask.multiprocessing.get):
    %time dd.read_csv('[path]/dask_to_csv.*.csv').set_index('my_index')

> CPU times: user 18.2 s, sys: 4.9 s, total: 23.1 s
> Wall time: 43 s
```

File: dask/bytes/core.py [hunk 1]
```diff
@@ -197,15 +197,34 @@ def read_bytes(urlpath, delimiter=None, not_zero=False, blocksize=2**27,
...
+        for path, offset, length, machine in zip(paths, offsets, lengths, machines):
+            with myopen(path, 'rb') as f:
+                header = read_block(f, offset[0], nbytes, delimiter, number_of_delimiters=1)
+                #beginning of file: read two lines (head + first data row)            
+                samples.append(read_block(f, offset[0], nbytes, delimiter, number_of_delimiters=2))
+                #block within file: append one data row to known header
+                samples_dummy = [header + read_block(f, o, nbytes, delimiter, number_of_delimiters=1)
+                     for o in offset[1:]]
+                samples.extend(samples_dummy)
+        return samples, out
```
Comment (mrocklin) [inline dask/bytes/core.py]: This code looks like it is specific to CSV files rather than reading bytes generally.

Comment (abast) [inline dask/bytes/core.py]: I see. Would something like the following be acceptable: We add a `samplefun=` keyword to the `read_bytes` function. `samplefun` will then be called with a (file, offset) tuple for each block. It is then up to samplefun to take care of a format specific sampeling.

**Comment by mrocklin** [comment]
Closing this as stale.  Please feel free to speak up if anyone would like to take this on again.

**Pull request closed without merging**